Mathlib/Combinatorics/Enumerative/DoubleCounting.lean

```diff
@@ -68,7 +68,13 @@ theorem mem_bipartiteAbove {b : β} : b ∈ t.bipartiteAbove r a ↔ b ∈ t ∧
 
 theorem sum_card_bipartiteAbove_eq_sum_card_bipartiteBelow [∀ a b, Decidable (r a b)] :
     (∑ a ∈ s, (t.bipartiteAbove r a).card) = ∑ b ∈ t, (s.bipartiteBelow r b).card := by
-  simp_rw [card_eq_sum_ones, bipartiteAbove, bipartiteBelow, sum_filter]
+   simp_rw [card_eq_sum_ones, sum_sum_bipartiteAbove_eq_sum_sum_bipartiteBelow]
+
+@[to_additive]
+theorem prod_prod_bipartiteAbove_eq_prod_prod_bipartiteBelow
+    [AddCommMonoid R] (f : α → β → R) [∀ a b, Decidable (r a b)] :
+    ∑ a ∈ s, ∑ b ∈ t.bipartiteAbove r a, f a b = ∑ b ∈ t, ∑ a ∈ s.bipartiteBelow r b, f a b := by
+  simp_rw [bipartiteAbove, bipartiteBelow, sum_filter]
```

Comment:
You need to swap the two theorems

Reply:
Swapped them and it works. But can you give me a crash course on how `@[to_additive]` works and why we use it here? The theorem name `prod_prod_bipartiteAbove_eq_prod_prod_bipartiteBelow` implies we're sing a product, but we use sums, but then magically we also have `sum_sum_bipartiteAbove_eq_sum_sum_bipartiteBelow`? Is it just that `AddCommMonoid` is written multiplicatively? But then why do we not need to use products in the theorem statement instead of sums?

Reply:
`to_additive` is a metaprogram that turns multiplicatively-written statements into additively-written ones. It also guesses the names automatically (but sometimes gets them wrong)